Cookies with SameSite=Strict should not be sent if a redirect is cross-site

TingPing opened the pull request:
#### 676141739a2842debf124ddfe7f5b49cf179be86
<pre>
Cookies with SameSite=Strict should not be sent if a redirect is cross-site
<a href="https://bugs.webkit.org/show_bug.cgi?id=272986">https://bugs.webkit.org/show_bug.cgi?id=272986</a>

Reviewed by Matthew Finkel.

As the spec states <a href="https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis#section-5.2">https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis#section-5.2</a>

&gt; The request is not the result of a cross-site redirect. That is, the origin of every url
&gt; in the request&apos;s url list is same-site with the request&apos;s current url&apos;s origin.

So when a cross-site redirect happens we set the request as not same-site.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/about-blank-nested.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/about-blank-subresource.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/about-blank-toplevel.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/fetch.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/form-get-blank-reload.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/form-get-blank.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/form-post-blank-reload.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/form-post-blank.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/get_all_cookies-default-samesite-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/get_named_cookie-default-samesite-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/iframe-reload.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/iframe.document.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/iframe.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/img.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/multiple-samesite-attributes.https-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/sandbox-iframe-nested.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/sandbox-iframe-subresource.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/setcookie-lax.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/setcookie-navigation.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/window-open-reload.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/window-open.https-expected.txt: Added.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::updateRequestAfterRedirection):

Canonical link: <a href="https://commits.webkit.org/278078@main">https://commits.webkit.org/278078@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/271bf3d62bc9e324a774b756671fd1ba1e5bc9a9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48453 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27665 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51412 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51141 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50759 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33601 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25187 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39610 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25343 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41827 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20730 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22826 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43001 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6510 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44768 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43479 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53046 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23500 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19826 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46924 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24765 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42025 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45843 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25570 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7104 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24488 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/multiple-samesite-attributes.https-expected.txt:
```diff
@@ -1,14 +1,42 @@
+Blocked access to external URL https://www1.localhost:9443/cookies/resources/redirectWithCORSHeaders.py?status=307&location=https%3A%2F%2F127.0.0.1%3A9443%2Fcookies%2Fresources%2FimgIfMatch.py%3Fname%3Dsamesite_unsupported%26value%3D0.10277761549984665
 
 PASS Same-host images are strictly same-site
-FAIL Subdomain images are strictly same-site promise_test: Unhandled rejection with value: "'samesite_unsupported_none=0.9461741832459153' not present on https://www1.web-platform.test:9443"
```
Any idea why these reverted from web-platform.test to 127.0.0.1?

TingPing replied:
This stood out to me also and I'm not sure why, but I'll try to look at it in more detail.

FYI the GLib ports run these tests with web-platform.test typically. The macOS runners do not, so that is one reason the test results here may not align with wpt.fyi.

sysrqb reviewed (approved):
Looks good to me. Note, the Apple ports do support the same-site attribute, but the implementation is in another library. I'll need to look at what is causing the test failures in our testing infrastructure for these at some point. They are partially passing upstream (https://wpt.fyi/results/cookies/samesite/fetch.https.html?label=experimental&label=master&aligned).

TingPing committed:
Cookies with SameSite=Strict should not be sent if a redirect is cross-site
https://bugs.webkit.org/show_bug.cgi?id=272986

Reviewed by Matthew Finkel.

As the spec states https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis#section-5.2

> The request is not the result of a cross-site redirect. That is, the origin of every url
> in the request's url list is same-site with the request's current url's origin.

So when a cross-site redirect happens we set the request as not same-site.

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/about-blank-nested.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/about-blank-subresource.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/about-blank-toplevel.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/fetch.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/form-get-blank-reload.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/form-get-blank.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/form-post-blank-reload.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/form-post-blank.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/get_all_cookies-default-samesite-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/get_named_cookie-default-samesite-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/iframe-reload.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/iframe.document.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/iframe.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/img.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/multiple-samesite-attributes.https-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/sandbox-iframe-nested.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/sandbox-iframe-subresource.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/setcookie-lax.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/setcookie-navigation.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/window-open-reload.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite/window-open.https-expected.txt: Added.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::updateRequestAfterRedirection):

Canonical link: https://commits.webkit.org/278078@main
